app/controllers/users_controller.rb

```diff
@@ -13,6 +13,59 @@ def show
     end
   end
 
+  def login
+    username = params[:user][:username]
+
+    user = User.find_by(username: username)
+    if user.nil?
+      flash_msg = "Welcome new user"
+    else
+      flash_msg = "Welcome back #{user.username}"
+    end
+
+    user ||= User.create(username: username)
+
+    session[:user_id] = user.id
+    flash[:success] = flash_msg
+    redirect_to users_path
+  end
+
+  def create
+    auth_hash = request.env["omniauth.auth"]
+
+    user = User.find_by(uid: auth_hash[:uid], provider: "github")
+    if user
+      flash[:success] = "Logged in as returning user #{user.username}"
+    else
+      user = User.build_from_github(auth_hash)
+
+      if user.save
+        flash[:success] = "Logged in as new user #{user.username}"
+      else
+        flash[:error] = "Could not create new user account: #{user.errors.messages}"
+        return redirect_to users_path
+      end
+    end
+
+    session[:user_id] = user.id
+    return redirect_to users_path
+  end
+
+  def current
+    @user = User.find_by(id: session[:user_id])
+    if @user.nil?
+      flash[:error] = "You must be logged in first!"
+      redirect_to users_path
+    end
+  end
+
+  def logout
```

Comment:
These look great!! I think the logout is defined as "destroy" in our route so we can change one or the other to match each other!